### PR TITLE
Verify the training artifact at the live evaluation precision

### PR DESCRIPTION
**Pull request by damenknight**
## What changed

`waldo model train` cannot complete for any model above roughly 76M parameters, on any GPU. Two commits, kept separate because they are separate bugs, but the first is why the second went unnoticed.

**Verify the training artifact at the live evaluation precision.** ADR 0044's check reloads `model.safetensors` and compares its held-out loss against the live number, failing outside the larger of 0.02 or one percent. The live number is measured under bfloat16 autocast (`record_evaluation` passes `mixed_precision=True`), while the reloaded model is cast to float32 and evaluated with `mixed_precision=False`. The comparison therefore measures autocast rounding rather than the reload. That error accumulates with depth, so it crosses the tolerance at a fixed model size and every larger model then fails verification for a fault it does not have.

The artifact is always *better* than the live model, which is the tell: fp32 evaluation beats bf16 evaluation of the same weights.

Artifact loss delta, same corpus and presets, on one AMD gfx1151 (ROCm 7.13) and one NVIDIA A30 (CUDA 12.8):

| preset | layers / width | before, ROCm | before, CUDA | after, ROCm | after, CUDA |
| --- | --- | --- | --- | --- | --- |
| 10m | 6 / 384 | -0.006949 | -0.006992 | +0.000101 | -0.000005 |
| 35m | 12 / 512 | -0.013628 | -0.013719 | +0.000000 | -0.000039 |
| 90m | 12 / 768 | **-0.021570 fails** | **-0.022532 fails** | +0.000115 | +0.000063 |

Two vendors, two drivers, two PyTorch builds, agreeing to under one percent, which is what you would expect if the cause is arithmetic rather than hardware. After the change the depth dependence is gone rather than merely reduced, and the live held-out losses are identical to six decimals, so training is untouched.

The residual is the bfloat16 weight round-trip the check exists to measure. The previous offset was large enough to hide it: a real serialization fault had to exceed 0.02 *on top of* a systematic -0.0216 before anyone saw it, so the check was failing good models and blind to bad ones at the same time. It now has roughly 170x headroom over its noise floor.

**Derive the expected worker revision in the lifecycle tests.** The MLX, PyTorch and TorchTitan lifecycle tests each assert a hardcoded backend revision of `r2`. Every one of those workers is at `r6`, so all three fail at that assertion on any machine able to run them. The suite still reports green because each skips wherever its runtime is absent and CI runs only `ingest-direct.sh`, which left the real training path with no working coverage.

They now read the revision from the constant the binary reports. Pinning the contract is worth keeping; pinning a *copy* of it is what went stale, and it would go stale again at the next bump — #1 already moves TorchTitan to `r7`.

## Verification

- `gofmt -l .` clean, `go vet ./...` clean, `./testing/all.sh` green on macOS.
- `testing/e2e/model-pytorch.sh` on Linux with real CUDA torch: **exits 1 on `main`** at the `r2` assertion, exits 0 here. Also re-run with the constant set to `r7` to confirm the derived assertion tracks a bump.
- The new guard in `internal/training` was checked against the unfixed worker to confirm it can fail, not just pass:
  ```
  --- FAIL: TestPyTorchWorkerEvaluatesArtifactAtLiveEvaluationPrecision
      held-out evaluations disagree on compute precision:
      "evaluate_model(self.model, mixed_precision=True)" vs
      "evaluate_model(artifact_model, mixed_precision=False)"
  ```
- Full training runs on both accelerators for the three presets above, before and after.

Two limitations worth stating rather than leaving to be found:

The guard is a source-level assertion, not a behavioural one. `model-pytorch.sh` trains a 1-layer, width-32 model, which is far too small to accumulate this error, so an assertion on its recorded delta would pass with or without the bug. That the only real-torch test is too small to catch depth-dependent faults is arguably the more interesting finding, and I did not want to ship a test that cannot fail.

`TestMLXBackendStreamsProtocolAndConsumesCompletion` is flaky under `-race`, failing 5 of 8 runs on **baseline `main`** and 7 of 8 here, with `mlx_test.go:77: MLX worker: read |0: file already closed`. It is unrelated to this change, which touches the PyTorch worker and shell scripts rather than the MLX path, and it appears intermittently in this repo's CI history. Flagging it so it is not misattributed.

## Contracts

No durable format changes and no ADR change. ADR 0044's stated intent is unaffected and better served: the check still verifies that persisted weights reload and retain measured quality, and now measures only that.

## DCO

- [x] Every commit includes a `Signed-off-by` trailer (`git commit --signoff`).

### CODE REVIEW

**Comment by gmkurtzer** [comment]
This looks good to me, thanks!
